File: c/lc/binary-search-tree-iterator.cc
class BSTIterator {
    stack<const TreeNode*> stack;
    
    void push_left_leaf(const TreeNode* root) {
        while (root) {
            stack.push(root);
            root = root->left;
        }
    }
public:
    BSTIterator(const TreeNode* root):stack() {
        push_left_leaf(root);
    }
    
    int next() {
        const TreeNode* top = stack.top();
        stack.pop();
        push_left_leaf(top->right);
        return top->val;
    }
    
    bool hasNext() {
        return !stack.empty();
    }
};

File: kotlin/src/dk/leetcode/BinarySearchTreeIterator.kt
package dk.leetcode
// https://leetcode.com/problems/binary-search-tree-iterator/

import java.util.*

class Solution_BSTIterator(root: TreeNode?) {

    val stack = Stack<TreeNode>()

    init {
        pushLeftBranch(root)
    }

    tailrec fun pushLeftBranch(root: TreeNode?) {
        if (root != null) {
            stack.add(root)
            pushLeftBranch(root.left)
        }
    }

    fun next(): Int {
        val top = stack.peek()
        stack.pop()
        pushLeftBranch(top.right)
        return top.`val`
    }

    fun hasNext() = stack.isNotEmpty()

}
